python/lsst/daf/butler/core/fileTemplates.py

```diff
@@ -0,0 +1,137 @@
+    """
+
+    def __init__(self, template):
+        self.template = template
+
+    def format(self, dataunits, datasettype=None, component=None):
+        """Format a template string into a full path.
+
+        Parameters
+        ----------
+        dataunits : `DataUnits`
+            DataUnits and the corresponding values.
+        datasettype : `str`, optional.
+            DataSetType name to use if needed. If it contains a "." separator
+            the type name will be split up into the main DataSetType and a
+            component.
+        component : `str`, optional
+            Component of a composite. If `datasettype` defines a component
+            this parameter will be ignored.
+
+        Returns
+        -------
+        path : `str`
+            Expanded path.
+
+        Raises
+        ------
+        KeyError
+            Requested field is not defined and the field is not optional.
+            Or, `component` is specified but "component" was not part of
+            the template.
+        """
+        units = dataunits.definedUnits()
+
+        if datasettype is not None:
+            # calexp.wcs means wcs component of a calexp
+            if "." in datasettype:
+                datasettype, component = datasettype.split(".", maxsplit=1)
+            units["datasettype"] = datasettype
+
+        usedComponent = False
+        if component is not None:
+            units["component"] = component
+
+        fmt = string.Formatter()
+        parts = fmt.parse(self.template)
+        output = ""
+
+        for literal, field_name, format_spec, conversion in parts:
+
+            if field_name == "component":
+                usedComponent = True
+
+            if "?" in format_spec:
+                optional = True
+                # Remove the non-standard character from the spec
+                format_spec = format_spec.replace("?", "")
+            else:
+                optional = False
+
+            if field_name in units:
+                value = units[field_name]
+            elif optional:
+                # If this is optional ignore the format spec
+                # and do not include the literal text prior to the optional
+                # field unless it contains a "/" path separator
+                format_spec = ""
+                value = ""
+                if "/" not in literal:
+                    literal = ""
+            else:
+                raise KeyError("{} requested in template but not defined and not optional".format(field_name))
+
+            # Now use standard formatting
+            output = output + literal + format(value, format_spec)
+
+        # Complain if we were meant to use a component
+        if component is not None and not usedComponent:
+            raise KeyError("Component {} specified but template {} did not use it".format(component,
+                                                                                          self.template))
```

Comment:
Why is this different from any other field?

Reply:
because if you explicitly specify a component but the component is not used that's going to be a problem because it means the template does not specify how to distinguish components.

Reply:
I was assuming that DataUnits could be far more detailed than is needed for the file templates. That's why I didn't check that every unit was used.